sup3r/utilities/loss_metrics.py

```diff
@@ -117,6 +117,103 @@ def __call__(self, x1, x2, sigma=1.0):
         return mmd
 
 
+class MaterialDerivativeLoss(tf.keras.losses.Loss):
+    """Loss class for the material derivative. This is the left hand side of
+    the Navier-Stokes equation and is equal to internal + external forces
+    divided by density.
+    https://en.wikipedia.org/wiki/Material_derivative
+    """
+
+    MAE_LOSS = MeanAbsoluteError()
+
+    def _derivative(self, x, axis=1):
+        """Custom derivative function for compatibility with tensorflow
+
+        Parameters
+        ----------
+        x : tf.Tensor
+            (n_observations, spatial_1, spatial_2, temporal)
+        axis : int
+            Axis to take derivative over
+        """
+        if axis == 1:
+            return tf.concat([x[:, 1:2] - x[:, 0:1],
+                              (x[:, 2:] - x[:, :-2]) / 2,
+                              x[:, -1:] - x[:, -2:-1]], axis=1)
+        elif axis == 2:
+            return tf.concat([x[..., 1:2, :] - x[..., 0:1, :],
+                              (x[..., 2:, :] - x[..., :-2, :]) / 2,
+                              x[..., -1:, :] - x[..., -2:-1, :]], axis=2)
+        elif axis == 3:
+            return tf.concat([x[..., 1:2] - x[..., 0:1],
+                              (x[..., 2:] - x[..., :-2]) / 2,
+                              x[..., -1:] - x[..., -2:-1]], axis=3)
```

Comment:
ignorant question: why are you doing a `[2:]` offset to calculate the derivative instead of just doing a 1-pixel difference?

Reply:
Gotcha that makes sense, maybe consider just adding this note for posterity?

Reply:
Yeah good idea